### ISSUE TITLE
Re-expressing executors built on work queue interfaces as `OneWayExecutor`s

### ISSUE THREAD

**Issue by brycelelbach** [executors/executors#417]
I've been doing some brainstorming about how we might move to one execution function and retain support for system-level work queue interfaces.

System-level work queue interfaces are analogous to system-level thread interfaces (for example OS thread interfaces like pthreads):

• Low level interface, part of the OS or system software stack.
• Main abstraction: work queues, which describes sequences of work to be executed in order. Analogous to OS threads.
• Often times, work queues are a software abstraction that virtualizes a fixed number of hardware work queues. Analogous to OS threads abstracting hardware cores.

CUDA is an example or system-level work queue interfaces. However, this design is not specific to GPUs or accelerators. There are a number of CPU based work queue systems - they can be common, for example, in network hardware and networking libraries.

Historically, we (@jaredhoberock/Agency) have implemented work queue executor and futures as follows:

• Work queue future == storage for content + a work queue.
• Work queue executor == target execution context.

In a purely one-way execution model, we would instead have:

• Sequenced work queue executor == target execution context + work queue (for example, device ordinal + stream in CUDA).
• Work queue future == storage for content + a work queue sequenced executor ("this executor", as opposed to a "next executor" bound by via).

This is necessary because with only one-way `execute`, we have no way of telling the executor which work queue it should use. This is the purpose of `then_execute` in P0443.

The solution I came up with is to make **be** the work queue.

A sequenced work queue executor, of course, executes all work submitted to it in order. It is analogous to a `std::thread` pool executor that has a single thread in it.

Obviously, this sequenced work executor is not the only executor you need, though.

How do we write a one-way work queue executor that executes all work independently?

Let's think about this in terms of another one-way executor that executes work independently: `new_std_thread_executor`, a `std::thread`-based one-way executor, which spawns and then detachs a `std::thread` everytime that `execute` is called:

```
struct new_std_thread_executor
{
  template <typename F>
  void execute(F f)
  {
    std::thread t(f);
    t.detach;
  }
};
```

We can write a similar new work queue executor:

```
struct new_work_queue_executor
{
  template <typename C>
  void execute(C c)
  {
    // Create a new work queue.
    sequenced_work_queue_executor new_e;
    // The workqueue is copied/moved into the lambda,
    // and thus kept alive.
    new_e.execute([=] { new_e.execute(c); });
  }
};
```

Alternatively, we may want to enqueue the lifetime management of the new work queue as separate item in the queue (this makes more sense for GPUs):

```
struct new_work_queue_executor
{
  template <typename C>
  void execute(C c)
  {
    myQueue s; // Create a new work queue.
    myWorkQueueCreate(&s);
    myEnqueue(s, c);
    // After `f`, decrement the reference count on the work queue.
    myEnqueue(s, [=] { myWorkQueueDecref(s); });
  }
};
```

In some frameworks, there are restrictions prohibiting you from calling framework APIs from within callbacks. These restrictions are typically to prevent you from calling APIs that block on work to complete, which would deadlock because you'd be blocking waiting on yourself. However, in this case we are just running cleanup code, so we should be fine.

These `new_*_executor`s are very naive, because both `std::thread`s and system-level work queue APIs are typically heavyweight virtualizations of a hardware resource. An "OS thread per future" or "work queue per future" isn't going to be very efficient. Also, if you create many OS threads/system-level work queues, they'll all have to share the same hardware resources. 

For example, on Linux, creating 1000 OS threads on an 8 core machine is unwise. Likewise, in CUDA, creating 1000 streams (e.g. work queues) is unwise, because NVIDIA GPUs usually have just 8 hardware channels (e.g hardware work queues).

Should we be able to express them? Yes, of course. But they aren't the best tool.

How about a `std_thread_pool_executor` and a `work_queue_pool_executor`? This makes more sense, and is likely to be a performant option.  

Now here's a tricky question. If one of my work queue futures contains a sequenced executor representing the work queue that the work it is associated with is in, how do I get that executor when I am using my new work queue executor or my work queue pool executor?

Ideally, I would want:

```
my_work_queue_pool_executor queue;
auto f = async(queue, c);
// f contains a `my_sequenced_executor` to the particular stream it was placed on.
```

I'm not sure how to do this without specializing async, because I need a way to determine which stream in the pool the work is running on. I don't have any way to get that information out of one way execute.

This would probably work by specializing `async`:

```
template <typename C>
auto async(my_work_queue_pool_executor exec, C c)
{
  using T = decltype(c());
  my_shared_state s;
  my_sequenced_executor this_e = exec.schedule([s] { s.value(c()); });
  return cuda_future<T>{this_e, s};
}
```

Note the use of a custom execution function `schedule`. It would be just like one-way, except it returns an executor for the particular work queue in the pool that the work was scheduled on.

Alternatively, if `execute` accepted functions that take an executor, as #413 suggests, I might be able to make things work. I'm not sure about that, though, because the function runs in the new execution agent, so it may not be possible to pass the executor to it.

I'm not sure what the solution to this issue is. I'm pretty close to a working solution for writing work queue executors as one-way executors and building futures on top of that, but this is a roadblock right now.

**Comment by LeeHowes** [comment]
There are now so many issues open here on very closely related topics that I no longer know where to comment.

However, what you are getting to here is very close to what we discussed at the weekend and what I then went away to implement.

So here's a PR against pushmi: https://github.com/facebookresearch/pushmi/pull/32/files (which we probably won't merge as-is because it is incomplete and does not define a clean finalised set of customisation points).

In this PR I've implemented a mechanism for taking a synchronous future chain - which I define as one where the type being passed is the actual value, and hence the next entry in the chain cannot start until the previous one completes, and forking it into an asynchronous chain and some conceptual out of band work. This is `async_fork`, parameterised by an executor.

`async_fork(e)` takes a value-based chain and converts it, based on a customisation point on `e` into a chain that passes an asynchronous completion token, that contains whatever state is meaningful to `e`, and a handle to the value, and passes that down the chain. In the simplest example of a new_thread executor this might include a condition variable that a thread can block on.

There is a corresponding join operation `async_join` which is not explicitly parameterised but is internally specialised on the incoming token, converting a series of out of band operations into a value chain again. This might, for example, start a thread that waits on a CV and when the value is ready, passes it down the chain.

Then I implemented `async_transform` and `async_bulk` where `async_bulk` is the value-only bulk roughly as in P0443 (with a slight modification to the set of functions called). Both of these functions also contain customisation point calls that are customised on the token being passed down the chain. 

I wrote examples in there with an inline_executor and a new_thread executor. Bulk is only implemented inline because I added that quickly yesterday. For inline this runs much as it would with normal value passing, running immediately when the previous future completes. For new_thread the customisation point behaves differently - for each node in the graph it launches a thread, which first blocks on the incoming CV and then runs the work and satisfies the outgoing CV. So all these threads are sitting waiting to process work and dispatched by a primary control flow path that runs straight through the chain to the `async_join` . Note that this customisation is hidden inside the operations.

Bulk could clearly be customised similarly - new_thread could launch one thread to run the loop. It could launch an array of threads to run the loop and join on them.

A GPU version of this would similarly customise on the executor, and instead of passing a condition_variable it would pass a handle to a queue, each task then appending a series of operations on the queue and where the join at the end would add the appropriate callback that would copy the result value back into a normal work chain.

It is important to note that none of this required any changes to Kirk's basic concepts. It is highly extensible, and yet also customisable. As we add new task creators, we can add default implementations and then allow each executor author to write a set of customisations for whichever of the standard task creators we have without any need for us to modify complex executor concepts in the future.

Note in particularly what this means about always-present error handling. For errors that reach async_fork we can keep them in value mode and bypass the entire asynchronous enqueue. For errors that occur queue-side we can propagate them in the token, which obviously means device-side code has to be able to handle that, or we can abort or whatever other mechanism is cleanest. There is no concern about switching to a GPU executor requiring that GPU tasks be able to handle incoming exceptions, that becomes a mechanism of the host side code of the GPU executor, not the asynchronous GPU-side work.

**Comment by LeeHowes** [comment]
Importantly here I want to separate bulkness from the asynchronous queuing behaviour. Bulk can be synchronous: an openmp executor might actually be entirely synchronous and not pass any special state, just changing my inline bulk code to do a `#pragma parallel for` - it would need the customisation point only because that would change the way the code is compiled, the data flow would be the same as this inline example.

While queuing work to the GPU is valid also for scalar code - there is no reason why scalar transform shouldn't also have a GPU executor specialisation to allow enqueue of a small task (or even a callback) that does some modification to the data before continuing.
